### PR TITLE
Ewelink discover fix and enhancements

### PR DESCRIPTION
**Pull request by NickDub-old**
### Pull Request check-list

To ensure your Pull Request can be accepted as fast as possible, make sure to review and check all of these items:

- [x] If your changes affects code, did your write the tests?
- [x] Are tests passing? (`npm test` on both front/server)
- [x] Is the linter passing? (`npm run eslint` on both front/server)
- [x] Did you run prettier? (`npm run prettier` on both front/server)
- [ ] If you are adding a new features/services, did you run integration comparator? (`npm run compare-translations` on front)
- [ ] If your changes modify the API (REST or Node.js), did you modify the API documentation? (Documentation is based on comments in code)
- [ ] If you are adding a new features/services which needs explanation, did you modify the user documentation? See [the GitHub repo](https://github.com/GladysAssistant/v4-website) and the [website](https://gladysassistant.com).
- [ ] Did you add fake requests data for the demo mode (`front/src/config/demo.json`) so that the demo website is working without a backend? (if needed) See [https://demo.gladysassistant.com](https://demo.gladysassistant.com).

NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open.

### Description of change

Please provide a description of the change here. It's always best with screenshots, so don't hesitate to add some!

### CODE REVIEW

File: server/services/ewelink/lib/device/poll.js [hunk 1]
```diff
@@ -1,13 +1,73 @@
...
+  if (currentTemperature && feature.last_value !== currentTemperature) {
+    logger.debug(`eWeLink: Polling device "${deviceId}", temperature new value = ${currentTemperature}`);
+    return currentTemperature;
+  }
+  return null;
+};
+
+// const pollEnergyPower = async (gladys, eWeLinkDevice, feature) => {
+//   const { deviceId } = parseExternalId(feature.external_id);
+//   const response = await connection.getDevicePowerUsage(deviceId);
```
Comment (atrovato) [inline server/services/ewelink/lib/device/poll.js]: Commented code is not required.

File: server/services/ewelink/lib/device/poll.js [hunk 2]
```diff
@@ -17,98 +77,61 @@ async function poll(device) {
...
-        state: currentBinaryState,
-      });
+  await Promise.mapSeries(device.features || [], async (feature) => {
+    let state;
+    switch (feature.category) {
+      case DEVICE_FEATURE_CATEGORIES.SWITCH: // Power
+        if (feature.type === DEVICE_FEATURE_TYPES.SWITCH.BINARY) {
+          state = pollPower(eWeLinkDevice, feature);
+        }
+        // } else if (feature.type === DEVICE_FEATURE_TYPES.SWITCH.POWER) {
```
Comment (atrovato) [inline server/services/ewelink/lib/device/poll.js]: Commented code is not required.

File: server/services/ewelink/lib/features/energyPower.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+const {
+  DEVICE_FEATURE_CATEGORIES,
+  DEVICE_FEATURE_TYPES,
+  DEVICE_FEATURE_UNITS,
+} = require('../../../../utils/constants');
+
+module.exports = {
+  // Gladys feature
+  generateFeature: (channel = 0) => {
```
Comment (atrovato) [inline server/services/ewelink/lib/features/energyPower.js]: Do we really generate power features? Power code looks fully commented.

File: server/services/ewelink/lib/features/index.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+const temperatureFeature = require('./temperature');
+
+const AVAILABLE_FEATURE_MODELS = {
+  power: {
+    uiid: [1, 2, 3, 4, 5, 6, 7, 8, 9, 14, 15],
+    feature: powerFeature,
+  },
+  // energyPower: {
+  //   uiid: [5],
+  //   feature: energyPowerFeature,
```
Comment (atrovato) [inline server/services/ewelink/lib/features/index.js]: Energy commented, commented code is not required.

**Review by atrovato — COMMENTED** [review]
Looks fine to me (code only), didn't try in real mode.

I really like the fact you upgrade the eWelink service package version :) 

One question, my device is offline, why can't I add it to Gladys?

image

**Comment by NickDub-old** [comment]
> One question, my device is offline, why can't I add it to Gladys?

If the device is offline, we cannot detect the available features and therefore, for the moment, we cannot add it.
In the future, I will add the ability to add and update an unhandled device.

File: server/services/ewelink/lib/device/poll.js (hunk 1 above)
Comment (NickDub-old) [reply server/services/ewelink/lib/device/poll.js]: Removed...

File: server/services/ewelink/lib/device/poll.js (hunk 2 above)
Comment (NickDub-old) [reply server/services/ewelink/lib/device/poll.js]: Removed...

File: server/services/ewelink/lib/features/energyPower.js (hunk 3 above)
Comment (NickDub-old) [reply server/services/ewelink/lib/features/energyPower.js]: Removed...

File: server/services/ewelink/lib/features/index.js (hunk 4 above)
Comment (NickDub-old) [reply server/services/ewelink/lib/features/index.js]: Removed...

**Review by atrovato — DISMISSED** [review]
Looks good to me, but I don't have any ewelink device anymore, maybe @VonOx have some to test it in real life.

**Comment by Pierre-Gilles** [comment]
@VonOx could you try this ? :) Looks good to me too, looking for a real world review before merging

**Comment by atrovato** [comment]
@VonOx do you still have eWelink devices?

**Comment by VonOx** [comment]
Yes sorry guys for delay, will test this week end.

**Comment by VonOx** [comment]
Some testing ( With sonoff basic ):

* Account setup OK
* Device discover ok
* Device added to dashboard ok ( but initial state was wrong , displayed on but was off )
* Switch to off , nothing happen ( ok because initial state was wrong )
* Turn on ( OK )
* Turn off ( NOK ) - Displayed off in dashboard but device is on
* After waiting device poll, real state is displayed in dashboard
* Turn off OK

Maybe polling related problem or maybe I'm too fast while testing
PR Seem's ok

EDIT: Got this error from nowhere after testing  ( ❤️ ITEAD server )
```
2021-03-12T17:34:37+0100 <error> device.poll.js:23 (DeviceManager.poll) There was an error while polling device ewelink-1000fb237d
2021-03-12T17:34:37+0100 <error> device.poll.js:24 (DeviceManager.poll) Error403: eWeLink: Authentication failed
    at EweLinkHandler.throwErrorIfNeeded (/home/vonox/repos/GladysEwilink/server/services/ewelink/lib/device/index.js:57:13)
    at EweLinkHandler.poll (/home/vonox/repos/GladysEwilink/server/services/ewelink/lib/device/poll.js:68:14)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async DeviceManager.poll (/home/vonox/repos/GladysEwilink/server/lib/device/device.poll.js:21:5) {
  status: 403,
  code: 'FORBIDDEN'
}
```

**Comment by Pierre-Gilles** [comment]
Thanks for your testing @VonOx  :) 

> Device added to dashboard ok ( but initial state was wrong , displayed on but was off )
> Switch to off , nothing happen ( ok because initial state was wrong )

So it seems there is an initial state problem? So PR is ok or not ? ^^

**Comment by NickDub-old** [comment]
The pr is OK, but there are vulnerabilities.
Can you check it ?

**Review by atrovato — DISMISSED** [review]
Ok looks good to me.

File: server/services/ewelink/lib/device/index.js [hunk 5]
```diff
@@ -57,7 +57,7 @@ async function throwErrorIfNeeded(response, emit = false, config = false) {
       throw new Error403(`eWeLink: ${response.msg}`);
     }
     if (emit) {
-      this.gladys.event.emit(EVENTS.WEBSOCKET.SEND_ALL, {
+      await this.gladys.event.emit(EVENTS.WEBSOCKET.SEND_ALL, {
```
Comment (Pierre-Gilles) [inline server/services/ewelink/lib/device/index.js]: this.gladys.event.emit is a synchronous function, no need for await

File: server/services/ewelink/lib/device/index.js [hunk 6]
```diff
@@ -41,7 +41,7 @@ async function throwErrorIfNeeded(response, emit = false, config = false) {
       this.accessToken = '';
       this.apiKey = '';
       if (emit) {
-        this.gladys.event.emit(EVENTS.WEBSOCKET.SEND_ALL, {
+        await this.gladys.event.emit(EVENTS.WEBSOCKET.SEND_ALL, {
```
Comment (Pierre-Gilles) [inline server/services/ewelink/lib/device/index.js]: this.gladys.event.emit is a synchronous function, no need for await

File: server/services/ewelink/lib/device/discover.js [hunk 7]
```diff
@@ -25,41 +27,28 @@ async function discover() {
...
-      const uiid = discoveredDevice.uiid.toString();
-      if (Object.keys(models).includes(uiid)) {
-        // ...if the model is supported, ...
-        const channels = await connection.getDeviceChannelCount(discoveredDevice.deviceid);
-        await this.throwErrorIfNeeded(channels, true);
+    await Promise.map(
+      discoveredDevices,
+      async (discoveredDevice) => {
+        // ...if it is already in Gladys...
+        const deviceInGladys = await this.gladys.stateManager.get(
```
Comment (Pierre-Gilles) [inline server/services/ewelink/lib/device/discover.js]: this.gladys.stateManager.get is a synchronous function, no need for await

**Review by Pierre-Gilles — CHANGES_REQUESTED** [review]
Hi! Thanks for your PR :) 

I wrote a few very minor changes related to code only. 

On the functional side, was this PR tested properly ?

The bug reported by @VonOx in previous messages were fixed ? 

If it's all good, and it has been tested, I'm good to merge! 🙂

**Comment by Pierre-Gilles** [comment]
Hello! Just saw your changes, looks better! 

On the functional side, was this PR tested properly ?

The bug reported by @VonOx in previous messages were fixed ?

**Comment by NickDub-old** [comment]
> The bug reported by @VonOx in previous messages were fixed ?

Not yet. I'm trying to reproduce it...
Otherwise, everything is OK.

**Comment by Pierre-Gilles** [comment]
Hello! Just wanted to know the status of this PR ? :)

If you need help, don't hesitate to ping me on the forum.

**Comment by NickDub-old** [comment]
I don't have any more ewelink devices to test, but if it's ok for you, this PR is ready to be merged...

**Comment by VonOx** [comment]
I have one, can test if you need.

**Comment by Pierre-Gilles** [comment]
OK @VonOx Let me know if the bug you previously saw is indeed fixed :)

**Comment by VonOx** [comment]
@Pierre-Gilles Not reproduced , so yes it's fixed

**Comment by Pierre-Gilles** [comment]
Do you confirm that this PR is retro-compatible with existing connected ewelink device ?

There is no breaking change, like users don't need to re-connect their device to make them work ? 

External_id used did not change?

**Comment by NickDub-old** [comment]
There is no change to the External_id.
This PR is retro-compatible with existing connected ewelink device ;)

**Review by Pierre-Gilles — APPROVED** [review]
Thanks for the PR then! :) 

I'm merging this !

**Pull request merged**